README.md

# charter

A collection of chart implementations for Flutter.

So far this supports the following:
 - PolarAreaChart

## Screenshots

![polar-area-chart-example.png](https://github.com/londonappstudio/flutter-charts/raw/master/screenshots/polar-area-chart-example.png)

## QuickStart

### Add Dependency
	dependencies:
		charter: 0.0.1-alpha-2

### Update Packages
	$ flutter pub get

### Import Package

	import 'package:charter/charter.dart';


Add the following code snippet and tweak

	PolarAreaChart.basic(
	  grid: [5,10,15], // Scale for the whole chart
	  features: [
	    "broccoli",
	    "cheese",
	    "salmon",
	    "potato",
	    "carrot",
	    "rice",
	    "lentils",
	    "covfefe",
	    "pasta",
	    "beef",
	    "grains"
	  ],
	  // Features and data must be same length
	  data: [3,1,5,14,6,2,9,13,5,6,10],
	  featuresTextStyle: TextStyle (
	    fontWeight: FontWeight.bold,
	    fontSize: 12
	  ),
	  drawSegmentDividers: true,
	  // Colours are repeated if less than data.length
	  featureColors: [
	    Colors.green,
	    Colors.blue,
	    Colors.red,
	    Colors.orange,
	    Colors.yellow,
	    Colors.pink,
	    Colors.brown
	  ]
	)



CHANGELOG.md
## [0.0.1-alpha-3] - 2020-05-07

* Accept Color objects instead of just MaterialColor

## [0.0.1-alpha-2] - 2020-05-07

* Added missing dispose() and update text x coord to take into account newlines

## [0.0.1-alpha-1] - 2020-05-07

* Added Polar Area Chart


test/fluttercharts_test.dart
import 'package:charter/flutter_test.dart';

import 'package:charter/charter.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
    expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  });
}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:londonappstudio/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/charter.dart
library flutter_charts;

import 'dart:ui';
import 'dart:math' as math;
import 'dart:developer';
import 'package:flutter/foundation.dart';

import 'package:flutter/material.dart';
import 'dart:math' show pi, cos, sin, max;

const defaultColours = [
  Colors.green,
  Colors.blue,
  Colors.red,
  Colors.orange,
  Colors.purple,
  Colors.yellow,
  Colors.brown
];

class PolarAreaChart extends StatefulWidget {
  final List<int> grid;
  final List<String> features;
  final List<double> data;
  final TextStyle ticksTextStyle;
  final TextStyle featureLabelsTextStyle;
  final bool drawSegmentDividers;
  final List<Color> featureColors;

  const PolarAreaChart({
    Key key,
    @required this.grid,
    @required this.features,
    @required this.data,
    this.ticksTextStyle = const TextStyle(color: Colors.grey, fontSize: 12),
    this.featureLabelsTextStyle = const TextStyle(color: Colors.black, fontSize: 16),
    this.drawSegmentDividers = false,
    this.featureColors = defaultColours,
  }) : super(key: key);

  factory PolarAreaChart.basic({
    @required List<int> grid,
    @required List<String> features,
    @required List<double> data,
    bool drawSegmentDividers = false,
    TextStyle featuresTextStyle,
    List<Color> featureColors = defaultColours
  }) {
    return PolarAreaChart(
      grid: grid,
      features: features,
      data: data,
      featureLabelsTextStyle: featuresTextStyle,
      drawSegmentDividers: drawSegmentDividers,
      featureColors: featureColors,
    );
  }

  @override
  _PolarAreaChartState createState() => _PolarAreaChartState();

}


class _PolarAreaChartState extends State<PolarAreaChart>
    with SingleTickerProviderStateMixin {
  double fraction;
  Animation<double> animation;
  AnimationController animationController;

  @override
  void initState() {
    super.initState();
    animationController = AnimationController(
        duration: Duration(milliseconds: 1000), vsync: this);

    animation = Tween(begin: 0.0, end: 1.0).animate(CurvedAnimation(
      curve: Curves.fastOutSlowIn,
      parent: animationController,
    ))
      ..addListener(() {
        setState(() {
          fraction = animation.value;
        });
      });

    animationController.forward();
  }

  @override
  void dispose() {
    animationController.dispose();
    super.dispose();
  }

  @override
  void didUpdateWidget(PolarAreaChart oldWidget) {
    super.didUpdateWidget(oldWidget);

    animationController.reset();
    animationController.forward();
  }

  @override
  Widget build(BuildContext context) {
    return CustomPaint(
      size: Size(double.infinity, double.infinity),
      painter: PolarAreaChartPainter(
          widget.grid,
          widget.features,
          widget.data,
          widget.ticksTextStyle,
          widget.featureLabelsTextStyle,
          widget.drawSegmentDividers,
          widget.featureColors,
          this.fraction),
    );
  }
}

class PolarAreaChartPainter extends CustomPainter {
  final List<int> gridLines;
  final List<String> features;
  final List<double> values;
  final TextStyle ticksTextStyle;
  final TextStyle featuresTextStyle;
  final List<Color> featureColors;
  final double fraction;
  final bool drawSegmentDividers;

  PolarAreaChartPainter(
      this.gridLines,
      this.features,
      this.values,
      this.ticksTextStyle,
      this.featuresTextStyle,
      this.drawSegmentDividers,
      this.featureColors,
      this.fraction);

  @override
  void paint(Canvas canvas, Size size) {

    //Calculate center point of canvas
    final centerX = size.width / 2.0;
    final centerY = size.height / 2.0;
    final centerOffset = Offset(centerX, centerY);

    //Radius of the graph has to be smaller than the
    final radius = math.min(centerX, centerY) * 0.70;
    final int maxValue = gridLines.last;

    //To determine how big to render the segments on the canvas
    final scale = radius / maxValue * 2;

    // Remember we are in Radians
    var angle = (2 * pi) / features.length;
    var segmentSize = angle;

    // Painting the chart outline (A circle)
    for(int i = 0; i < features.length; i++){

      Path outline = Path();
      var outlinePaint = Paint()
        ..color = featureColors[i%featureColors.length].withOpacity(0.6)
        ..style = PaintingStyle.stroke
        ..strokeWidth = 5.0
        ..isAntiAlias = true;

      outline.moveTo(centerX, centerY);
      outline.arcTo(
          Rect.fromCenter(
            center: centerOffset,
            width: scale*maxValue,
            height: scale*maxValue
          ),
          angle*i,
          angle,
          true
      );
      outline.moveTo(centerX, centerY);
      outline.close();
      canvas.drawPath(outline, outlinePaint);
    }


    // Paint for segmentDividers and grid scale
    var gridLinePaint = Paint()
      ..color = Colors.grey.withOpacity(0.5)
      ..style = PaintingStyle.stroke
      ..strokeWidth = 1.0
      ..isAntiAlias = true;


    // Painting the circles and labels for the given ticks (could be auto-generated)
    // The last tick is ignored, since it overlaps with the feature label
    var tickDistance = radius / (gridLines.length);
    var tickLabels = gridLines;

    tickLabels.sublist(0, gridLines.length - 1).asMap().forEach((index, tick) {
      var tickRadius = tickDistance * (index + 1);

      canvas.drawCircle(centerOffset, tickRadius, gridLinePaint);

      TextPainter(
        text: TextSpan(text: tick.toString(), style: ticksTextStyle),
        textDirection: TextDirection.ltr,
      )
        ..layout(minWidth: 0, maxWidth: size.width)
        ..paint(canvas,
            Offset(centerX, centerY - tickRadius - ticksTextStyle.fontSize));
    });


    if(drawSegmentDividers){
      features.asMap().forEach((index, feature) {
        var xAngle = cos((angle * index));
        var yAngle = sin((angle * index));
        var featureOffset = Offset(centerX + radius * xAngle, centerY + radius * yAngle);
        canvas.drawLine(centerOffset, featureOffset, gridLinePaint);
      });
    }

    features.asMap().forEach((index, feature) {

      var xAngle = cos((angle * index) + (angle / 2));
      var yAngle = sin((angle * index) + (angle / 2));

      var labelRadius = radius * 1.1;

      var featureOffset = Offset(
          centerX + labelRadius * xAngle,
          centerY + labelRadius * yAngle
      );

      var maxLineLength = feature.split("\n").map((e) => e.length).reduce((value, element) => math.max(value, element));
      var featureLabelFontHeight = featuresTextStyle.fontSize;
      var featureLabelFontWidth = featuresTextStyle.fontSize - 4;
      var labelYOffset = yAngle < 0 ? - featureLabelFontHeight : 0;
      var labelXOffset = xAngle < 0 ?
              (featureLabelFontWidth * maxLineLength * xAngle) : // Left side
              - (featureLabelFontWidth * maxLineLength * (1 - xAngle)) / 2; // Right side

      var coloredTextStyle = featuresTextStyle.copyWith(color: featureColors[index%featureColors.length]);
      TextPainter(
        text: TextSpan(text: feature,
              style: coloredTextStyle
        ),
        textAlign: TextAlign.center,
        textDirection: TextDirection.ltr,
      )
        ..layout(minWidth: 0, maxWidth: size.width)
        ..paint(
            canvas,
            Offset(featureOffset.dx + labelXOffset, featureOffset.dy + labelYOffset));
    });

    // Paint the slices
    for (var i = 0; i < values.length; i++) {

      var slicePaint = Paint()
        ..color = featureColors[i%featureColors.length].withOpacity(0.6)
        ..style = PaintingStyle.fill;

      var path = Path();
      path.moveTo(centerX, centerY);
      path.arcTo(
          Rect.fromCenter(
              center: centerOffset,
              width: scale * values[i],
              height: scale * values[i]
          ),
          i*segmentSize,
          segmentSize,
          false);

      path.lineTo(centerX, centerY);
      path.close();
      canvas.drawPath(path, slicePaint);
    }

  }

  @override
  bool shouldRepaint(PolarAreaChartPainter oldDelegate) {
    return oldDelegate.fraction != fraction;
  }
}

lib/fluttercharts.dart
library flutter_charts;

import 'dart:ui';
import 'dart:math' as math;

import 'package:flutter/material.dart';
import 'dart:math' show pi, cos, sin, max;

const defaultColours = [
  Colors.green,
  Colors.blue,
  Colors.red,
  Colors.orange,
  Colors.purple,
  Colors.yellow,
  Colors.brown
];

class PolarAreaChart extends StatefulWidget {
  final List<int> ticks;
  final List<String> features;
  final List<int> data;
  final bool reverseAxis;
  final TextStyle ticksTextStyle;
  final TextStyle featuresTextStyle;
  final Color outlineColor;
  final Color axisColor;
  final List<Color> graphColors;

  const PolarAreaChart({
    Key key,
    @required this.ticks,
    @required this.features,
    @required this.data,
    @required this.reverseAxis,
    this.ticksTextStyle = const TextStyle(color: Colors.grey, fontSize: 12),
    this.featuresTextStyle = const TextStyle(color: Colors.black, fontSize: 16),
    this.outlineColor = Colors.black,
    this.axisColor = Colors.grey,
    this.graphColors = defaultColours,
  }) : super(key: key);

  factory PolarAreaChart.light({
    @required List<int> ticks,
    @required List<String> features,
    @required List<int> data,
    bool reverseAxis = false,
  }) {
    return PolarAreaChart(
      ticks: ticks,
      features: features,
      data: data,
      reverseAxis: reverseAxis,
    );
  }

  @override
  _PolarAreaChartState createState() => _PolarAreaChartState();

}


class _PolarAreaChartState extends State<PolarAreaChart>
    with SingleTickerProviderStateMixin {
  double fraction;
  Animation<double> animation;
  AnimationController animationController;

  @override
  void initState() {
    super.initState();
    animationController = AnimationController(
        duration: Duration(milliseconds: 1000), vsync: this);

    animation = Tween(begin: 0.0, end: 1.0).animate(CurvedAnimation(
      curve: Curves.fastOutSlowIn,
      parent: animationController,
    ))
      ..addListener(() {
        setState(() {
          fraction = animation.value;
        });
      });

    animationController.forward();
  }

  @override
  void didUpdateWidget(PolarAreaChart oldWidget) {
    super.didUpdateWidget(oldWidget);

    animationController.reset();
    animationController.forward();
  }

  @override
  Widget build(BuildContext context) {
    return CustomPaint(
      size: Size(double.infinity, double.infinity),
      painter: PolarAreaChartPainter(
          widget.ticks,
          widget.features,
          widget.data,
          widget.reverseAxis,
          widget.ticksTextStyle,
          widget.featuresTextStyle,
          widget.outlineColor,
          widget.axisColor,
          widget.graphColors,
          this.fraction),
    );
  }
}

class PolarAreaChartPainter extends CustomPainter {
  final List<int> gridLines;
  final List<String> features;
  final List<int> values;
  final bool reverseAxis;
  final TextStyle ticksTextStyle;
  final TextStyle featuresTextStyle;
  final Color outlineColor;
  final Color axisColor;
  final List<Color> graphColors;
  final double fraction;

  PolarAreaChartPainter(
      this.gridLines,
      this.features,
      this.values,
      this.reverseAxis,
      this.ticksTextStyle,
      this.featuresTextStyle,
      this.outlineColor,
      this.axisColor,
      this.graphColors,
      this.fraction);

  @override
  void paint(Canvas canvas, Size size) {
    final centerX = size.width / 2.0;
    final centerY = size.height / 2.0;
    final centerOffset = Offset(centerX, centerY);
    final radius = math.min(centerX, centerY) * 0.70;
    final double maxValue = 15; // TODO: Needs to be configured or based on data
    final scale = radius / maxValue * 2;

    print("");

    var gridLines = [maxValue];


    // Take these as a param from somewhere


    num degToRad(num deg) => deg * (math.pi / 180.0);

    // Painting the axis for each given feature
    var segmentSize = 360 / values.length; // Degrees
    var angle = (2 * pi) / features.length; // Radians

    // Painting the chart outline (A circle)
    for(int i = 0; i < features.length; i++){

      Path outline = Path();
      var outlinePaint = Paint()
        ..color = graphColors[i%graphColors.length].withOpacity(0.6)
        ..style = PaintingStyle.stroke
        ..strokeWidth = 5.0
        ..isAntiAlias = true;

      outline.moveTo(centerX, centerY);
      outline.arcTo(
          Rect.fromCenter(
            center: centerOffset,
            width: scale*maxValue,
            height: scale*maxValue
          ),
          angle*i,
          angle,
          true
      );
      outline.moveTo(centerX, centerY);
      outline.close();
      canvas.drawPath(outline, outlinePaint);

//      canvas.drawCircle(centerOffset, radius, outlinePaint);
    }

    var ticksPaint = Paint()
      ..color = axisColor.withOpacity(0.5)
      ..style = PaintingStyle.stroke
      ..strokeWidth = 1.0
      ..isAntiAlias = true;

//    canvas.drawCircle(centerOffset, radius, outlinePaint);

    // Painting the circles and labels for the given ticks (could be auto-generated)
    // The last tick is ignored, since it overlaps with the feature label
    // Painting the circles and labels for the given ticks (could be auto-generated)
    // The last tick is ignored, since it overlaps with the feature label
    var tickDistance = radius / (gridLines.length);
    var tickLabels = reverseAxis ? gridLines.reversed.toList() : gridLines;

    tickLabels.sublist(0, gridLines.length - 1).asMap().forEach((index, tick) {
      var tickRadius = tickDistance * (index + 1);

      canvas.drawCircle(centerOffset, tickRadius, ticksPaint);

      TextPainter(
        text: TextSpan(text: tick.toString(), style: ticksTextStyle),
        textDirection: TextDirection.ltr,
      )
        ..layout(minWidth: 0, maxWidth: size.width)
        ..paint(canvas,
            Offset(centerX, centerY - tickRadius - ticksTextStyle.fontSize));
    });



    features.asMap().forEach((index, feature) {
      var xAngle = cos((angle * index) + (angle / 2));
      var yAngle = sin((angle * index) + (angle / 2));

      var featureOffset =
      Offset(centerX + radius * xAngle, centerY + radius * yAngle);

//      canvas.drawLine(centerOffset, featureOffset, ticksPaint);

      var featureLabelFontHeight = (featuresTextStyle as TextStyle).fontSize;
      var featureLabelFontWidth = (featuresTextStyle as TextStyle).fontSize - 4;
      var labelYOffset = yAngle < 0 ? - featureLabelFontHeight : 0;
      var labelXOffset =
      xAngle < 0 ? -featureLabelFontWidth * feature.length : 0;

      var coloredTextStyle = featuresTextStyle.copyWith(color: graphColors[index%graphColors.length]);

      TextPainter(
        text: TextSpan(text: feature, style: coloredTextStyle),
        textAlign: TextAlign.center,
        textDirection: TextDirection.ltr,
      )
        ..layout(minWidth: 0, maxWidth: size.width)
        ..paint(
            canvas,
            Offset(featureOffset.dx + labelXOffset,
                featureOffset.dy + labelYOffset));
    });

//    // Painting the axis for each given feature
//    var angle = (2 * pi) / features.length;
//
//    features.asMap().forEach((index, feature) {
//      var xAngle = cos(angle * index - pi / 2);
//      var yAngle = sin(angle * index - pi / 2);
//
//      var featureOffset =
//      Offset(centerX + radius * xAngle, centerY + radius * yAngle);
//
//      canvas.drawLine(centerOffset, featureOffset, ticksPaint);
//
//      var featureLabelFontHeight = (featuresTextStyle as TextStyle).fontSize;
//      var featureLabelFontWidth = (featuresTextStyle as TextStyle).fontSize - 4;
//      var labelYOffset = yAngle < 0 ? -featureLabelFontHeight : 0;
//      var labelXOffset =
//      xAngle < 0 ? -featureLabelFontWidth * feature.length : 0;
//
//      TextPainter(
//        text: TextSpan(text: feature, style: featuresTextStyle),
//        textAlign: TextAlign.center,
//        textDirection: TextDirection.ltr,
//      )
//        ..layout(minWidth: 0, maxWidth: size.width)
//        ..paint(
//            canvas,
//            Offset(featureOffset.dx + labelXOffset,
//                featureOffset.dy + labelYOffset));
//    });



    // Paint the slices
    for (var i = 0; i < values.length; i++) {

      var slicePaint = Paint()
        ..color = graphColors[i%defaultColours.length].withOpacity(0.6)
        ..style = PaintingStyle.fill;

      var edgesPaint = Paint()
        ..color = graphColors[i%defaultColours.length].withOpacity(0.6)
        ..style = PaintingStyle.stroke
        ..strokeWidth = 1.5
        ..isAntiAlias = true;

      var path = Path();
      path.moveTo(centerX, centerY); // OK
      path.arcTo(
          Rect.fromCenter(
              center: centerOffset,
              width: scale * values[i],
              height: scale * values[i]
          ),
          degToRad(i*segmentSize),
          degToRad(segmentSize),
          false);
      path.lineTo(centerX, centerY);
//          lastend+(Math.PI*2*(values[i]/myTotal)));


//      path.lineTo(100.0 + (i * 5), 200);


      path.close();
      canvas.drawPath(path, slicePaint); // fill it in
//      canvas.drawPath(path, edgesPaint);


      //Stolen from a JS fiddle so I don't have to do any math
//      ctx.beginPath();
//      ctx.moveTo(200,150);
//      ctx.arc(200,150,myRadius[i],lastend,lastend+
//          (Math.PI*2*(myData[i]/myTotal)),false);
//      console.log(myRadius[i]);
//      ctx.lineTo(200,150);
//      ctx.fill();
//      lastend += Math.PI*2*(myData[i]/myTotal);
    }


//
//    // Painting each graph
//    data.asMap().forEach((index, graph) {
//      var graphPaint = Paint()
//        ..color = graphColors[index % graphColors.length].withOpacity(0.3)
//        ..style = PaintingStyle.fill;
//
//      var graphOutlinePaint = Paint()
//        ..color = graphColors[index % graphColors.length]
//        ..style = PaintingStyle.stroke
//        ..strokeWidth = 2.0
//        ..isAntiAlias = true;
//
//      // Start the graph on the initial point
//      var scaledPoint = scale * graph[0] * fraction;
//      var path = Path();
//
//      if (reverseAxis) {
//        path.moveTo(centerX, centerY - (radius * fraction - scaledPoint));
//      } else {
//        path.moveTo(centerX, centerY - scaledPoint);
//      }
//
//      graph.asMap().forEach((index, point) {
//        if (index == 0) return;
//
//        var xAngle = cos(angle * index - pi / 2);
//        var yAngle = sin(angle * index - pi / 2);
//        var scaledPoint = scale * point * fraction;
//
//        if (reverseAxis) {
//          path.lineTo(centerX + (radius * fraction - scaledPoint) * xAngle,
//              centerY + (radius * fraction - scaledPoint) * yAngle);
//        } else {
//          path.lineTo(
//              centerX + scaledPoint * xAngle, centerY + scaledPoint * yAngle);
//        }
//      });
//
//      path.close();
//      canvas.drawPath(path, graphPaint);
//      canvas.drawPath(path, graphOutlinePaint);
//    });
  }

  @override
  bool shouldRepaint(PolarAreaChartPainter oldDelegate) {
    return oldDelegate.fraction != fraction;
  }
}

example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:charter/charter.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
        // This makes the visual density adapt to the platform that you run
        // the app on. For desktop platforms, the controls will be smaller and
        // closer together (more dense) than on mobile platforms.
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();

}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      // This call to setState tells the Flutter framework that something has
      // changed in this State, which causes it to rerun the build method below
      // so that the display can reflect the updated values. If we changed
      // _counter without calling setState(), then the build method would not be
      // called again, and so nothing would appear to happen.
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: Column(
          // Column is also a layout widget. It takes a list of children and
          // arranges them vertically. By default, it sizes itself to fit its
          // children horizontally, and tries to be as tall as its parent.
          //
          // Invoke "debug painting" (press "p" in the console, choose the
          // "Toggle Debug Paint" action from the Flutter Inspector in Android
          // Studio, or the "Toggle Debug Paint" command in Visual Studio Code)
          // to see the wireframe for each widget.
          //
          // Column has various properties to control how it sizes itself and
          // how it positions its children. Here we use mainAxisAlignment to
          // center the children vertically; the main axis here is the vertical
          // axis because Columns are vertical (the cross axis would be
          // horizontal).
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Expanded(
                child: PolarAreaChart.basic(
                  grid: [5,10,15], // Scale for the whole chart
                  features: [
                    "broccoli",
                    "cheese",
                    "salmon",
                    "potato",
                    "carrot",
                    "rice\ngammonamon",
                    "lentils",
                    "covfefe",
                    "pasta",
                    "beef",
                    "grains"
                  ],
                  data: [3,1,5,14,6.5,2,9.8,13,5,6,10],
                  featuresTextStyle: TextStyle (
                    fontWeight: FontWeight.bold,
                    fontSize: 12
                  ),
                  drawSegmentDividers: true,
                  featureColors: [
                    Colors.green,
                    Colors.blue,
                    Colors.red,
                    Colors.orange,
                    Colors.yellow,
                    Colors.pink,
                    Colors.brown,
                    Color(0xff374ABE),
                    Colors.red[600]
                  ]
                )
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.headline4,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
